kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayerWithUpdate.java

```diff
@@ -0,0 +1,156 @@
+    private int currentBuildInfoNum = 0;
+
+    public void submit(LayerJobEntity job) {
+        completionService.submit(new Callable<LayerJobResult>() {
+            @Override
+            public LayerJobResult call() throws Exception {
+                KylinConfig.setAndUnsetThreadLocalConfig(job.config);
+                Thread.currentThread().setName("thread-build-layer" + job.info.getLayoutId());
+                try {
+                    return job.build();
+                } catch (Throwable t) {
+                    logger.error("Error occurred when run " + job.info.getLayoutId(), t);
+                    return new LayerJobResult( t);
+                }
+            }
+        });
+        currentBuildInfoNum ++;
+    }
+
+    public void updateLayer() {
+        for (int i = 0; i < currentBuildInfoNum; i++) {
+            try {
+                logger.info("Wait to take layer result.");
+                LayerJobResult result = completionService.take().get();
+                logger.info("Take job layer successful.");
+                if (result.isFailed()) {
+                    shutDownPool();
+                    throw new RuntimeException(result.getThrowable());
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                shutDownPool();
+                throw new RuntimeException(e);
+            }
+        }
+        currentBuildInfoNum = 0;
+    }
+
+    private void shutDownPool() {
+        pool.shutdown();
+        try {
+            pool.awaitTermination(10, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            logger.warn("Error occurred when shutdown thread pool.", e);
+            pool.shutdownNow();
+        }
+    }
+
+    private static class LayerJobResult {
+        private Throwable throwable;
+
+        LayerJobResult(Throwable throwable) {
+            this.throwable = throwable;
+        }
+
+        boolean isFailed() {
+            return throwable != null;
+        }
+
+        Throwable getThrowable() {
+            return throwable;
+        }
+
+    }
+
+    public static class LayerJobEntity {
+
+        NBuildSourceInfo info;
+        SpanningTree st;
+        SegmentInfo seg;
+        CubeBuildJob job;
+        KylinConfig config;
+        Map<Long, Long> cuboidsRowCount;
+
+        public LayerJobEntity(NBuildSourceInfo info, SpanningTree st, KylinConfig config,
+                              CubeBuildJob job, SegmentInfo seg, Map<Long, Long> cuboidsRowCount){
+            this.info = info;
+            this.config = config;
+            this.job = job;
+            this.st = st;
+            this.seg = seg;
+            this.cuboidsRowCount = cuboidsRowCount;
+        }
+
+        public LayerJobResult build() {
+            Collection<LayoutEntity> toBuildCuboids = info.getToBuildCuboids();
+            Dataset<Row> parentDS = info.getParentDS();
+            if (toBuildCuboids.size() > 1) {
+                parentDS.cache();
```

Comment:
How to set other storage level? It's better to add a configuration to support setting other storage level.